### ISSUE TITLE
Performance of maml

### ISSUE THREAD

**Issue by sudarshan1994** [oscarknagg/few-shot#4]

**Comment by sudarshan1994** [comment]
Hey thanks for your implementation of all three models. I found it very useful. I used matching networks implementation of yours in my research. Thank you for that. But when I run maml, I am not able to get more than 25 percent on 1 shot 5 way even after 35 epochs. Am I supposed to do some changes to the code before running it ? Also the training loss does not  change after the 3rd epoch. Please let me know if some minor edits are required to be done from my side

**Comment by sudarshan1994** [comment]
One aother difference is that a folder called logs is created, which by default has two sub directories called matching_nets and proto_nets but doesnot have one for maml, and maml.py throws an error which is fixed when we create a folder called maml inside logs

**Comment by sudarshan1994** [comment]
Any help would be appreciated a lot !!

**Comment by oscarknagg** [comment]
You're correct about the folder issue. I'll update the `setup_dirs()` util function to resolve that.

For your other issue I'll need a few more pieces of information.
- What arguments are you giving to `maml.py`?
- What version of PyTorch are you using?

I'll need to try replicate this myself in order to debug.

**Comment by sudarshan1994** [comment]
Hey, thank you so much for your response
1) I just ran it with the default setting python maml.py --data miniImageNet

2) pyotrch version is 1

Thanks again for getting back, please let  me know if you have further questions

**Comment by sudarshan1994** [comment]
One additional info : it seems to work on omniglot, just doesnt seem to miniImageNet

**Comment by sudarshan1994** [comment]
Hey were you able to replicate the bug ? or is it just that my setup is weird

**Comment by oscarknagg** [comment]
Hey, sorry for the slow follow-up I've been quite busy recently. I'm still able to get good performance with MAML on miniImageNet. I ran the exact hyperparams given in the original paper and get the following results.

```
(venv) oscar@oscar:~/PycharmProjects/few-shot$ pip freeze | grep torch
torch==1.0.1.post2
torchvision==0.2.1
(venv) oscar@oscar:~/PycharmProjects/few-shot$ python -m experiments.maml --dataset miniImageNet --order 1 --n 1 --k 5 --q 15 --meta-batch-size 4     --inner-train-steps 5 --inner-val-steps 10 --inner-lr 0.01 --eval-batches 40 --epoch-len 100
miniImageNet_order=1_n=1_k=5_metabatch=4_train_steps=5_val_steps=10
Indexing background...
48000it [00:00, 359754.61it/s]
Indexing evaluation...
12000it [00:00, 448608.65it/s]
Training MAML on miniImageNet...
Begin training...
Epoch 1: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 100/100 [03:34<00:00, 22.72s/it, loss=1.57, categorical_accuracy=0.292]
etc..
Epoch 50: 100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 100/100 [03:41<00:00, 23.38s/it, loss=1.38, categorical_accuracy=0.422]
Finished.
```

Training curves look like this. Performance is already close to the reported numbers after only 1/3 of the training used in the paper and we aren't close to convergence yet.
![image](https://user-images.githubusercontent.com/20032640/56428227-8b2c0a80-62b6-11e9-9e7a-207a0825fdc4.png)


Try running with exactly the parameters I did, perhaps the default ones in my script aren't good. If that doesn't work then maybe your miniImageNet data is messed up somehow. This seems especially likely as you still get good performance on Omniglot.

**Comment by sudarshan1994** [comment]
Thank you so much for your help, I assumed that maml.py was already configured to the optimal hyper parameters, went through checking other parts of your code, but not the hyperparams. This is super helpful for my research.

Begin training...                                                                                                        │        
Epoch 1: 100%|██████████████████████████████████| 100/100 [03:38<00:00, 22.77s/it, loss=1.57, categorical_accuracy=0.297]

btw, categorical_accuracy=0.297 is the average of the training accuracy and the validation accuracy right ? It is not just the validation  accuracy right ?

The categorical accuracy we calculate in the "on_epoch end " function is the validation accuracy correct ??

Thanks a lot for your help
